### PR TITLE
VisualFactory replaces basically everything

### PR DESCRIPTION
**Pull request by rjrjr**
_This is nowhere near ready for review. It's not my fault that GitHub insists on bothering people about draft PRs._

@helios175 prototyped a drastically stripped down alternative to ScreenViewFactory and all its relatives. This branch applies that model to Workflow, particularly with an eye toward fixing #546.

* `interface VisualFactory<ContextT, RenderingT, VisibleT>` replaces all of ViewRegistry, ScreenViewFactoryFinder, ScreenViewFactory, ComposeScreenViewFactory OverlayDialogFactory, etc.
  * `VisibleT` might be `View`, `@Composable (Modifier) -> Unit`, `Dialog`...
  * `ContextT` is `ContextOrContainer` for `View`, `Context` for `Dialog` factories, `Unit` for Compose
  * There are no subtypes of `VisualFactory`, which makes it trivial to decorate

* `VisualFactory` builds `VisualHolder`, which replaces `ScreenViewHolder`, etc.

* `VisualFactoryConverter` simplifies converting a factory for one type of visual for another, e.g. to wrap `@Composable fun Content()` factories with `ComposeView`, or in the other direction to wrap a `View` builder with `@Composable fun AndroidView()`

* `VisualEnvironment` is just a new name for `ViewEnvironment`, moved to a new package. A `ViewEnvironment` wrapper still exists in the old location, but I don't know if this will really be practical for migration

* `Visualizer` provides the core support for `WorkflowViewStub`, `@Composable WorkflowRendering()`, other containers. 
  * Also replaces the static `ScreenViewFactory.startShowing` and `ScreenViewHolder.show` functions, I'm pretty sure.
  * The plan is to have one `Visualizer` able to turn any rendering into a `View`, and another able to turn any rendering into a Composable. Feature developers will bind their leaf `View` and composable factories to specific `VisualEnvironmentKey`s, and the `Visualizers` will wrap them with appropriate converters, standard factories, etc.

Most test are broken, but known passing tests are:

* `WorkflowViewStubLifecycleTest` exercising both `VisualFactory` implementations and a `VisualFactory` based wrapper of `ScreenViewFactoryFinder` for backward compatibility.

* `EnvironmentScreenAndroidIntegrationTest`

Lots and lots of `TODO` comments that should become github issues.

Biggest bit of remaining uncertainty is all the nooks and crannies of `OverlayDialogHolder`, in support of

* back button hooks
* bounds management
* modal behavior

Latest thinking there is to make `VisualT` into a struct that wraps the `Dialog` and its `updateBounds` method. For back button concerns, we may be able to get out of that business entirely with the recent introduction of `ViewTreeOnBackPressedDispatcherOwner` to JetPack. We should address that ASAP on the main branch.

Also still need to finish demonstrating the Compose integration, there is some WIP, but the prototype @helios175 built gives me a lot of confidence there.

### CODE REVIEW

File: workflow-ui/core-common/src/main/java/com/squareup/workflow1/visual/ExactTypeVisualFactory.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        )?.let { delegateHolder ->
+          object : VisualHolder<RenderingT, VisualT> by delegateHolder {
+            override fun update(rendering: RenderingT): Boolean {
+              return predicate(rendering) && delegateHolder.update(rendering)
+            }
+          }
+        }
+      } else null
+    }
+  }
```
Comment (helios175) [inline workflow-ui/core-common/src/main/java/com/squareup/workflow1/visual/ExactTypeVisualFactory.kt:54]: Can move to a different file.

File: workflow-ui/core-common/src/main/java/com/squareup/workflow1/visual/MultiRendering.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * different types or compatibilities, and so require different factories and holders.
+ */
+@WorkflowUiExperimentalApi
+public abstract class MultiRendering<UiContextT, VisualT> {
+
+  private var currentHolder: VisualHolder<Any, VisualT>? = null
+
+  public val currentVisual: VisualT get() = currentHolder!!.visual
+
+  public val currentVisualOrNull: VisualT? get() = currentHolder?.visual
```
Comment (helios175) [inline workflow-ui/core-common/src/main/java/com/squareup/workflow1/visual/MultiRendering.kt:16]: I'd aim to name them without `current`. No strong opinion.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt [hunk 3]
```diff
@@ -210,36 +213,30 @@ public class WorkflowViewStub @JvmOverloads constructor(
...
+      WorkflowLifecycleOwner.installOn(newView)
+      doFirstUpdate()
+
+      if (inflatedId != NO_ID) newView.id = inflatedId
+      if (updatesVisibility) newView.visibility = visibility
+      background?.let { newView.background = it }
+      propagateSavedStateRegistryOwner(newView)
+
+      replaceOldViewInParent(parent, newView)
+    }
```
Comment (helios175) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt:239]: I'd eventually aim for making replaceWith receive a `doUpdate = true / false` and use false for this case.
Or have onNewVisualBeforeUpdate / onNewVisualAfterUpdate. 🤷 
Having MultiRendering do the update for us makes sense, but it shouldn't complicate us.

Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt:239]: Not really following. I _love_ this, providing a lambda that gets called only if there's an update to do. Seems much simpler than what I was doing before.

Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowViewStub.kt:239]: Maybe give the lambda three arguments, build `oldViewOrNull` into it.

**Comment by rjrjr** [comment]
This is a much cleaner API than the thing we evolved, but moving to it would require a herculean effort that we just can't make. It is also not the case that this is required to fix #546, I have something simpler in the works for that.

Still interested in landing this some day possibly at its own library, maybe Compose-first, but it's not part of the Workflow roadmap

**Pull request closed without merging**